renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -117,10 +118,211 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
   return false;
 }
 
+bool IsShaderParameterVisible(DXBC::ShaderType shaderType, D3D12_SHADER_VISIBILITY shaderVisibility)
+{
+  if(shaderVisibility == D3D12_SHADER_VISIBILITY_ALL)
+    return true;
+
+  if(shaderType == DXBC::ShaderType::Vertex && shaderVisibility == D3D12_SHADER_VISIBILITY_VERTEX)
+    return true;
+
+  if(shaderType == DXBC::ShaderType::Pixel && shaderVisibility == D3D12_SHADER_VISIBILITY_PIXEL)
+    return true;
+
+  return false;
+}
+
 void D3D12DebugManager::CreateShaderGlobalState(ShaderDebug::GlobalState &global,
                                                 DXBC::DXBCContainer *dxbc)
 {
-  RDCUNIMPLEMENTED("CreateShaderGlobalState not yet implemented for D3D12");
+  D3D12RenderState &rs = m_pDevice->GetQueue()->GetCommandData()->m_RenderState;
+  D3D12ResourceManager *rm = m_pDevice->GetResourceManager();
+
+  // Get the root signature
+  const D3D12RenderState::RootSignature *pRootSignature = NULL;
+  if(dxbc->m_Type == DXBC::ShaderType::Compute && rs.compute.rootsig != ResourceId())
+  {
+    pRootSignature = &rs.compute;
+  }
+  else if(rs.graphics.rootsig != ResourceId())
+  {
+    pRootSignature = &rs.graphics;
+  }
+
+  if(pRootSignature)
+  {
+    WrappedID3D12RootSignature *pD3D12RootSig =
+        rm->GetCurrentAs<WrappedID3D12RootSignature>(pRootSignature->rootsig);
+
+    size_t numParams = RDCMIN(pD3D12RootSig->sig.params.size(), pRootSignature->sigelems.size());
+    for(size_t i = 0; i < numParams; ++i)
+    {
+      const D3D12RootSignatureParameter &param = pD3D12RootSig->sig.params[i];
+      const D3D12RenderState::SignatureElement &element = pRootSignature->sigelems[i];
+      if(IsShaderParameterVisible(dxbc->m_Type, param.ShaderVisibility))
+      {
+        // Note that constant buffers are not handled as part of the shader global state
+
+        if(param.ParameterType == D3D12_ROOT_PARAMETER_TYPE_SRV && element.type == eRootSRV)
+        {
+          UINT shaderReg = param.Descriptor.ShaderRegister;
+          ID3D12Resource *pResource = rm->GetCurrentAs<ID3D12Resource>(element.id);
+          D3D12_RESOURCE_DESC resDesc = pResource->GetDesc();
+
+          ShaderDebug::FillViewFmt(DXGI_FORMAT_R32_UINT, global.srvs[shaderReg].format);
```

Comment:
Can't root buffers be any 32-bit width type? So UINT/SINT/FLOAT? This might all just fall out in the wash with memcpy's and DXBC's typeless registers, but if so it'd be good to leave a comment here just to say that.

Reply:
I added a TODO to check later if the D3D12_RESOURCE_DESC has the correct typed format. I'd need to build a test case to find out.